File: features/update/basic/feature.ts
import { WorkflowUpdateFailedError } from '@temporalio/client';
import { Feature } from '@temporalio/harness';
import * as wf from '@temporalio/workflow';
import * as assert from 'assert';

const WORKFLOW_INITIAL_STATE = '';
const UPDATE_ARG = 'update-arg';
const BAD_UPDATE_ARG = 'reject-me';
const UPDATE_RESULT = 'update-result';
const UPDATE_RESULT_NOT_YET_RECEIVED = 'update-result-not-yet-received';

const myUpdate = wf.defineUpdate<string, [string]>('myUpdate');

/**
 * A workflow with a signal and signal validator. If accepted, the signal makes
 * a change to workflow state. The workflow does not terminate until such a
 * change occurs.
 */
export async function workflow(): Promise<string> {
  let state = WORKFLOW_INITIAL_STATE;
  const update = (arg: string) => {
    state = arg;
    return UPDATE_RESULT;
  };
  const validate = (arg: string) => {
    if (arg == BAD_UPDATE_ARG) {
      throw new Error('Invalid Update argument');
    }
  };
  wf.setHandler(myUpdate, update, validate);
  await wf.condition(() => state != WORKFLOW_INITIAL_STATE);
  return state;
}

export const feature = new Feature({
  workflow,
  checkResult: async (runner, handle) => {
    var badUpdateResult = UPDATE_RESULT_NOT_YET_RECEIVED;
    try {
      badUpdateResult = await handle.executeUpdate(myUpdate, { args: [BAD_UPDATE_ARG] });
    } catch (err) {
      if (!(err instanceof WorkflowUpdateFailedError)) {
        throw err;
      }
    }
    assert.equal(badUpdateResult, UPDATE_RESULT_NOT_YET_RECEIVED);

    const updateResult = await handle.executeUpdate(myUpdate, { args: [UPDATE_ARG] });
    assert.equal(updateResult, UPDATE_RESULT);
    const workflowResult = await runner.waitForRunResult(handle);
    assert.equal(workflowResult, UPDATE_ARG);
  },
});

File: features/update/basic_async/feature.py
from datetime import timedelta

from temporalio import workflow
from temporalio.client import WorkflowHandle, WorkflowUpdateFailedError

from harness.python.feature import Runner, register_feature

WORKFLOW_INITIAL_STATE = ""
UPDATE_ARG = "update-arg"
BAD_UPDATE_ARG = "reject-me"
UPDATE_RESULT = "update-result"
UPDATE_RESULT_NOT_YET_RECEIVED = "update-result-not-yet-received"


@workflow.defn
class Workflow:
    """
    A workflow with a signal and signal validator.
    If accepted, the signal makes a change to workflow state.
    The workflow does not terminate until such a change occurs.
    """

    def __init__(self) -> None:
        self._state = WORKFLOW_INITIAL_STATE

    @workflow.run
    async def run(self) -> str:
        await workflow.wait_condition(lambda: self._state != WORKFLOW_INITIAL_STATE)
        return self._state

    @workflow.update
    async def my_update(self, arg: str) -> str:
        self._state = arg
        return UPDATE_RESULT

    @my_update.validator
    def my_validate(self, arg: str):
        if arg == BAD_UPDATE_ARG:
            raise ValueError("Invalid Update argument")


async def checker(_: Runner, handle: WorkflowHandle):
    bad_update_handle = await handle.start_update("my_update", BAD_UPDATE_ARG)
    bad_update_result = UPDATE_RESULT_NOT_YET_RECEIVED
    try:
        bad_update_result = await bad_update_handle.result()
    except WorkflowUpdateFailedError:
        pass
    else:
        assert False, "Expected Update to be rejected due to validation failure"
    assert bad_update_result == UPDATE_RESULT_NOT_YET_RECEIVED

    update_handle = await handle.start_update("my_update", UPDATE_ARG)
    update_result = await update_handle.result()
    assert update_result == UPDATE_RESULT
    result = await handle.result()
    assert result == UPDATE_ARG


register_feature(workflows=[Workflow], check_result=checker)
